### PR TITLE
Crop new picture

### LINKED ISSUE #488: Allow users to crop and zoom profile images
Use libraries like react-easy-crop or other to implement the following:

1. allow users to zoom out and crop profile images on the /profile/[userId] page.
2. the resulting image should be always 300 x 300 px square
3. on all of /profile/[userId], /mentors, and /mentors/[mentorId] page, display the resulting image (and not the original image) using CSS without requiring a different image URL (ie. the user shouldn't need to re-upload the image).

### PR DESCRIPTION
**Pull request by ada-only-one-ada**
Fixed #488

### CODE REVIEW

File: src/pages/profiles/[userId].tsx [hunk 1]
```diff
@@ -204,6 +205,8 @@ function Picture({ userId, profile, updateProfile, SaveButton }: {
   SaveButton: React.ComponentType,
 }) {
   const [me] = useUserContext();
+  const [isModalOpen, setModalOpen] = useState(false);
+  const [imageUrl, setImageUrl] = useState(profile.照片链接);
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:209]: Don't use this extra variable. Read image url directly from the profile parameter.

File: src/pages/profiles/[userId].tsx [hunk 2]
```diff
@@ -215,6 +217,10 @@ function Picture({ userId, profile, updateProfile, SaveButton }: {
     [userId, profile]
   );
 
+  const handleUpdateProfileImage = (x: number, y: number, zoom: number) => {
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:220]: `updateProfileImageParams`

File: src/pages/profiles/[userId].tsx [hunk 3]
```diff
@@ -200,10 +201,11 @@ function Basic({ user, profile, setUser, setProfile }: {
 function Picture({ userId, profile, updateProfile, SaveButton }: {
   userId: string,
   profile: UserProfile,
-  updateProfile: (k: keyof UserProfile, v: string) => void,
+  updateProfile: (k: keyof UserProfile, v: any) => void,
   SaveButton: React.ComponentType,
 }) {
   const [me] = useUserContext();
+  const [isModalOpen, setModalOpen] = useState(false);
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:208]: `isCropImageModalOpen`

File: src/pages/profiles/[userId].tsx [hunk 4]
```diff
@@ -246,7 +259,7 @@ function Picture({ userId, profile, updateProfile, SaveButton }: {
           可见，用于在个别情况下直接引用其他网站的图像：</Text>
         </FormHelperTextWithMargin>
         <Input bg="white" value={profile.照片链接 || ""} mb={componentSpacing}
-          onChange={ev => updateProfile('照片链接', ev.target.value)}
+          onChange={ev => {updateProfile('照片链接', ev.target.value);}}
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:262]: remove unnecessary characters like `{` and `;`

File: src/shared/UserProfile.ts [hunk 5]
```diff
@@ -1,6 +1,11 @@
 import { z } from "zod";
 
 const zStr = z.string().optional();
+export const zPicParams = z.object({
```
Comment (weihanwang) [inline src/shared/UserProfile.ts:4]: `zImageParams`

File: src/shared/UserProfile.ts [hunk 6]
```diff
@@ -1,6 +1,11 @@
...
+  x: z.number(),
+  y: z.number(),
+  zoom: z.number(),
```
Comment (weihanwang) [inline src/shared/UserProfile.ts:7]: comment on their meanings

File: src/pages/profiles/[userId].tsx [hunk 7]
```diff
@@ -64,7 +65,7 @@ export default function Page() {
   const [profile, setProfile] = useState<UserProfile>();
   useEffect(() => setProfile(oldProfile), [oldProfile]);
 
-  const updateProfile = (k: keyof UserProfile, v: string) => {
+  const updateProfile = (k: keyof UserProfile, v: any) => {
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:68]: `string | ImageParams` instead. avoid `any`

File: src/shared/UserProfile.ts [hunk 8]
```diff
@@ -1,6 +1,11 @@
...
+}).optional();
```
Comment (weihanwang) [inline src/shared/UserProfile.ts:8]: remove `optional()` from here. add it to lin 26.

File: src/components/MarkdownStyler.tsx [hunk 9]
```diff
@@ -15,7 +16,7 @@ import rehypeRaw from 'rehype-raw';
  * @params allowHtml Whether to parse raw html tags in content. DANGEROUS!
  */
 export default function MarkdownStyler({ content, allowHtml }: {
-  content: string,
+  content: string | ImageParams,
```
Comment (weihanwang) [inline src/components/MarkdownStyler.tsx:19]: why do you need to change this file? we should not need to change this file for this feature. It doesn't make sense for MarkdownStyler to take ImageParams as content type.

Instead, use `context as string` at where MarkdownStyler is called.

File: src/components/UserCards.tsx [hunk 10]
```diff
@@ -133,8 +133,8 @@ function search(users: MinUserAndProfile[], searchTerm: string) {
   const lower = searchTerm.trim().toLowerCase();
   const pinyin = toPinyin(lower);
 
-  const match = (v: string | null | undefined) => {
-    if (!v) return false;
+  const match = (v: string | null | undefined | ImageParams) => {
```
Comment (weihanwang) [inline src/components/UserCards.tsx:136]: same here. you shouldn't need to change this function.

File: src/pages/profiles/[userId].tsx [hunk 11]
```diff
@@ -65,11 +67,18 @@ export default function Page() {
...
     const updated = structuredClone(profile);
-    if (v) updated[k] = v;
-    else delete updated[k];
+    if (!v) {
+      delete updated[k];
+    } else {
+      if (typeof v === 'string' && k !== '照片参数') {
+        updated[k] = v;
+      } else if (typeof v === 'object' && k === '照片参数') {
+        updated[k] = v;
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:79]: you should keep the original logic and only adjust type casting or simply bypass type checking for ling 71 in the original version.

File: src/pages/profiles/[userId].tsx [hunk 12]
```diff
@@ -252,10 +261,11 @@ function Basic({ user, profile, setUser, setProfile }: {
 function Picture({ userId, profile, updateProfile, SaveButton }: {
   userId: string,
   profile: UserProfile,
-  updateProfile: (k: keyof UserProfile, v: string) => void,
+  updateProfile: (k: keyof UserProfile, v: string | ImageParams) => void,
   SaveButton: React.ComponentType,
 }) {
   const [me] = useUserContext();
+  const [isCropImageModalOpen, setCropImageModalOpen] = useState(false);
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:268]: For simplicity, change to `isCropping` and `setIsCropping`

File: src/pages/profiles/[userId].tsx [hunk 13]
```diff
@@ -285,6 +299,14 @@ function Picture({ userId, profile, updateProfile, SaveButton }: {
             <HStack><MdCloudUpload /><Text>上传照片</Text></HStack>
           }
         </Link>
+        {profile.照片链接 && 
+          <HStack onClick={()=>setCropImageModalOpen(!isCropImageModalOpen)}>
```
Comment (weihanwang) [inline src/pages/profiles/[userId].tsx:303]: `onClick={() => setIsCropping(true)}`. also notice necessary spaces. Use Cursor to correct missing or extra spaces and other punctuations.

File: src/shared/UserProfile.ts [hunk 14]
```diff
@@ -2,6 +2,15 @@ import { z } from "zod";
...
 const zStr = z.string().optional();
+const zImageParams = z.object({
+  // x represents the horizontal position of the image
+  x: z.number(),
+  // y represents the vertical position of the image
+  y: z.number(),
+  // zoom represents the zoom level of the image
+  zoom: z.number(),
+});
+export type ImageParams = z.infer<typeof zImageParams>;
```
Comment (weihanwang) [inline src/shared/UserProfile.ts:13]: please follow existing convention: `z.TypeOf<typeof zImageParams>`

File: src/components/CropImageModal.tsx [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+import { useState } from 'react';
+import Cropper from 'react-easy-crop';
+
+export function CropImageModal({ onClose, imageUrl, onSave } : {
+  onClose: () => void,
+  imageUrl?: string,
+  onSave: (x: number, y: number, zoom: number) => void,
+ }) {
+  const [crop, setCrop] = useState<{x: number, y: number}>({ x: 0, y: 0 });
+  const [zoom, setZoom] = useState<number>(2);
```
Comment (weihanwang) [inline src/components/CropImageModal.tsx:22]: why 2? document any "magic number" in comments

File: src/components/CropImageModal.tsx [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+  const handleSave = () => {
```
Comment (weihanwang) [inline src/components/CropImageModal.tsx:24]: `save`

File: src/components/CropImageModal.tsx [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+  Box,
+} from "@chakra-ui/react";
+import ModalWithBackdrop from "./ModalWithBackdrop";
+import { componentSpacing } from "theme/metrics";
+import { useState } from 'react';
+import Cropper from 'react-easy-crop';
+
+export function CropImageModal({ onClose, imageUrl, onSave } : {
+  onClose: () => void,
+  imageUrl?: string,
```
Comment (weihanwang) [inline src/components/CropImageModal.tsx:18]: you shouldn't need `?` here

File: src/components/CropImageModal.tsx [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    onSave(crop.x, crop.y, zoom); 
+    onClose(); 
+  };
+
+  return <ModalWithBackdrop isOpen onClose={onClose}>
+    <ModalContent>
+      <ModalHeader>修改图像</ModalHeader>
+      <ModalCloseButton />
+      <ModalBody>
+        {imageUrl &&
```
Comment (weihanwang) [inline src/components/CropImageModal.tsx:34]: remove this line

File: src/components/CropImageModal.tsx [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+              onCropChange={setCrop}
+              onZoomChange={setZoom}
+              cropSize={{ width: 300, height: 300 }}
+            />
+          </Box>
+        }
+      </ModalBody>
+      <ModalFooter>
+        <Flex gap={componentSpacing}>
+          <Button onClick={handleSave}>保存</Button>
```
Comment (weihanwang) [inline src/components/CropImageModal.tsx:55]: `variant="brand"`

**Comment by ada-only-one-ada** [comment]
open another new pr to resolve this.

**Pull request closed without merging**